src/kdc/tgs_policy.c

```diff
@@ -252,19 +252,21 @@ check_tgs_s4u2proxy(kdc_realm_t *kdc_active_realm,
 }
 
 static int
-check_tgs_u2u(kdc_realm_t *kdc_active_realm,
-              krb5_kdc_req *req, const char **status)
+check_tgs_u2u(kdc_realm_t *kdc_active_realm, krb5_kdc_req *req,
+              krb5_const_principal server_princ, const char **status)
 {
+    krb5_const_principal second_server_princ;
+
     if (req->kdc_options & KDC_OPT_ENC_TKT_IN_SKEY) {
         /* Check that second ticket is in request. */
         if (!req->second_ticket || !req->second_ticket[0]) {
             *status = "NO_2ND_TKT";
             return KDC_ERR_BADOPTION;
         }
-        /* Check that second ticket is a TGT. */
-        if (!krb5_principal_compare(kdc_context,
-                                    req->second_ticket[0]->server,
-                                    tgs_server)) {
+        /* Check that second ticket is a TGT to the server realm. */
+        second_server_princ = req->second_ticket[0]->server;
+        if (!is_local_tgs_principal(second_server_princ) ||
```

Comment:
I'm curios what so special about local (root) tgts, say in this case what's wrong with a non-local tgt? The client can still be non-local, that is could ask for a "local" tgt, using its non-local one and proceed, so why local?

Reply:
I don't know if there's a security reason.  In user-to-user the second ticket is typically received from the other party (without its session key) before being presented to the KDC.  If the request is made using other party's realm, one would expect the other party to have simply acquired a local TGT.

Reply:
Ack, thanks for clarifying it.